### ISSUE TITLE
TypeError on provisioning virtual parameters

### ISSUE THREAD

**Issue by covin** [zaidka/genieacs#243]
Hello,

I'm running GenieACS at current HEAD of master. Without any virtual parameters everything works fine. But setting up even the simplest kind of virtual parameter breaks the provisioning.
```javascript
return {writable: false, value: ["foo", "xsd:string"]};
```

During provisioning genie-cwmp throws either:
```
2017-06-27T09:10:21.334Z [ERROR] Uncaught exception; exceptionName="TypeError" exceptionMessage="Cannot read property alias of undefined" exceptionStack="TypeError: Cannot read property alias of undefined
    at Object.addPathMeta (/opt/genieacs/lib/common.js:236:12)
    at /opt/genieacs/lib/session.js:1944:14
    at /opt/genieacs/lib/local-cache.js:339:5
    at /opt/genieacs/lib/local-cache.js:79:14
    at /opt/genieacs/lib/cache.js:174:16
    at handleCallback (/opt/genieacs/node_modules/mongodb/lib/utils.js:95:56)
    at /opt/genieacs/node_modules/mongodb/lib/collection.js:1406:5
    at handleCallback (/opt/genieacs/node_modules/mongodb/lib/utils.js:95:56)
    at /opt/genieacs/node_modules/mongodb/lib/cursor.js:674:5
    at handleCallback (/opt/genieacs/node_modules/mongodb-core/lib/cursor.js:171:5)"
```
using Mongo as caching engine or:
```
2017-06-27T09:27:41.241Z [ERROR] Uncaught exception; exceptionName="TypeError" exceptionMessage="Cannot read property alias of undefined" exceptionStack="TypeError: Cannot read property alias of undefined
    at Object.addPathMeta (/opt/genieacs/lib/common.js:236:12)
    at /opt/genieacs/lib/session.js:1944:14
    at /opt/genieacs/lib/local-cache.js:339:5
    at Command.callback (/opt/genieacs/lib/local-cache.js:79:14)
    at normal_reply (/opt/genieacs/node_modules/redis/index.js:721:21)
    at RedisClient.return_reply (/opt/genieacs/node_modules/redis/index.js:819:9)
    at JavascriptRedisParser.returnReply (/opt/genieacs/node_modules/redis/index.js:192:18)
    at JavascriptRedisParser.execute (/opt/genieacs/node_modules/redis-parser/lib/parser.js:574:12)
    at Socket.<anonymous> (/opt/genieacs/node_modules/redis/index.js:274:27)
    at emitOne (events.js:96:13)"
```
using Redis.

**Comment by akcoder** [comment]
Try going back one check in. I'm betting this is related to the most recent commit. 

-dan

**Comment by covin** [comment]
Same result on previous commits. One can reproduce the error on fresh install using a simple provision script containing a declare statement on a "real' parameter and the virtual parameter:
```
const now = Date.now();
declare("InternetGatewayDevice.DeviceInfo.*", {value: now});
declare("VirtualParameters.*", {value: now});
```
Using `now` here to enforce GetParameterValue RPC here. Inserting an explicit `commit()` just before VirtualParameter declaration circumvents the error.

**Comment by zaidka** [comment]
I couldn't reproduce the issue using your script. Try the latest commit and make sure to run "npm run compile". The problematic commit that @akcoder was referring to is now fixed.

**Comment by covin** [comment]
I'm already on your commit f444ff7 from Friday  ("optimize loadPath function"). I think the problem @akcoder had in mind was an error like

```
2017-07-03T08:37:44.014Z [ERROR] Uncaught exception; exceptionName="MongoError" exceptionMessage="Cannot update VirtualParameters.bar._timestamp and VirtualParameters.bar._timestamp at the same time" exceptionStack="MongoError: Cannot update VirtualParameters.bar._timestamp and VirtualParameters.bar._timestamp at the same time
    at Function.MongoError.create (/opt/genieacs/node_modules/mongodb-core/lib/error.js:31:11)
    at toError (/opt/genieacs/node_modules/mongodb/lib/utils.js:114:22)
    at /opt/genieacs/node_modules/mongodb/lib/collection.js:1049:67
    at /opt/genieacs/node_modules/mongodb-core/lib/connection/pool.js:455:18
    at _combinedTickCallback (internal/process/next_tick.js:67:7)
    at process._tickCallback (internal/process/next_tick.js:98:9)"
```
Right and thanks, this has been solved by 9dd6980. Hm, not sure why my issue is not reproducible? I've tested it several times.

Presets/Provisions/VirtualParameters from MongoDB:
```
MongoDB server version: 3.4.3
> db.presets.find({})
{ "_id" : "init", "channel" : "", "weight" : 0, "precondition" : "{\"_tags\":\"_init\"}", "configurations" : [ { "type" : "provision", "name" : "init" }, { "type" : "delete_tag", "tag" : "_init" } ], "schedule" : "", "events" : {  } }
> db.provisions.find({})
{ "_id" : "init", "script" : "const now = Date.now();\r\ndeclare(\"InternetGatewayDevice.DeviceInfo.*\", {value: now});\r\ndeclare(\"VirtualParameters.*\", {value: now});" }
> db.virtualParameters.find({})
{ "_id" : "foo", "script" : "return {writable: false, value: [\"foo\", \"xsd:string\"]};" }
```

Running genieacs-cwmp:
```
2017-07-03T08:55:45.898Z [INFO] GenieACS (genieacs-cwmp) starting; pid=85885 version="1.1.1" dependencies="node@6.9.5,later@1.2.0,libxmljs@0.18.4,mongodb@2.2.24,redis@2.7.1,seedrandom@2.4.2" config="CONFIG_DIR=/opt/genieacs/config,MONGODB_CONNECTION_URL=mongodb://genie:xxx@127.0.0.1/genieacs-devel,CWMP_INTERFACE=127.0.0.1,CWMP_PORT=7547,CWMP_SSL=false,NBI_INTERFACE=127.0.0.1,NBI_PORT=7557,FS_INTERFACE=0.0.0.0,FS_PORT=7567,FS_HOSTNAME=192.168.0.1,FS_IP=192.168.0.1,LOG_INFORMS=true,DEBUG=true,REDIS_PORT=6379,REDIS_HOST=,REDIS_DB=0,CWMP_WORKER_PROCESSES=0,CWMP_LOG_FILE=,CWMP_ACCESS_LOG_FILE=,NBI_WORKER_PROCESSES=0,NBI_SSL=false,NBI_LOG_FILE=,NBI_ACCESS_LOG_FILE=,FS_WORKER_PROCESSES=0,FS_SSL=false,FS_LOG_FILE=,FS_ACCESS_LOG_FILE=,UDP_CONNECTION_REQUEST_PORT=0,DOWNLOAD_TIMEOUT=3600,EXT_TIMEOUT=3000,MAX_CACHE_TTL=86400,RETRY_DELAY=300,SESSION_TIMEOUT=30,CONNECTION_REQUEST_TIMEOUT=2000,GPN_NEXT_LEVEL=0,GPV_BATCH_SIZE=32,MAX_DEPTH=16,LOG_FORMAT=simple,ACCESS_LOG_FORMAT=,XML_RECOVER=false,XML_IGNORE_ENC=false,XML_FORMAT=false,XML_NO_DECL=false,XML_NO_EMPTY=false,XML_IGNORE_NAMESPACE=false,DEVICE_ONLINE_THRESHOLD=4000"
2017-07-03T08:55:46.636Z [INFO] Worker listening; pid=85905 address="127.0.0.1" port=7547
2017-07-03T08:55:46.637Z [INFO] Worker listening; pid=85911 address="127.0.0.1" port=7547
2017-07-03T08:55:46.637Z [INFO] Worker listening; pid=85912 address="127.0.0.1" port=7547
2017-07-03T08:55:46.651Z [INFO] Worker listening; pid=85936 address="127.0.0.1" port=7547
2017-07-03T08:55:46.674Z [INFO] Worker listening; pid=85930 address="127.0.0.1" port=7547
2017-07-03T08:55:46.677Z [INFO] Worker listening; pid=85924 address="127.0.0.1" port=7547
2017-07-03T08:55:46.698Z [INFO] Worker listening; pid=85918 address="127.0.0.1" port=7547
2017-07-03T08:55:46.744Z [INFO] Worker listening; pid=85937 address="127.0.0.1" port=7547
2017-07-03T08:55:50.243Z [INFO] 127.0.0.1 00040E-FRITZ%21Box-...: Inform; cpeRequestId="8498" informEvent="2 PERIODIC" informRetryCount=1
2017-07-03T08:56:14.120Z [INFO] 127.0.0.1 00040E-FRITZ%21Box-...: Inform; cpeRequestId="8499" informEvent="2 PERIODIC" informRetryCount=0
2017-07-03T08:56:14.551Z [INFO] 127.0.0.1 00040E-FRITZ%21Box-...: ACS request; acsRequestId="15d07aa172d100" acsRequestName="GetParameterValues"
2017-07-03T08:56:15.014Z [ERROR] Uncaught exception; exceptionName="TypeError" exceptionMessage="Cannot read property alias of undefined" exceptionStack="TypeError: Cannot read property alias of undefined
    at Object.addPathMeta (/opt/genieacs/lib/common.js:236:12)
    at /opt/genieacs/lib/session.js:1950:14
    at /opt/genieacs/lib/local-cache.js:339:5
    at /opt/genieacs/lib/local-cache.js:218:15
    at handleCallback (/opt/genieacs/node_modules/mongodb/lib/utils.js:95:56)
    at /opt/genieacs/node_modules/mongodb/lib/collection.js:1051:5
    at /opt/genieacs/node_modules/mongodb-core/lib/connection/pool.js:455:18
    at _combinedTickCallback (internal/process/next_tick.js:67:7)
    at process._tickCallback (internal/process/next_tick.js:98:9)"
```

**Comment by zaidka** [comment]
Thanks, I managed to reproduce it. Will fix soon.
